proposals/downloads_get_file_hash.md

```diff
@@ -0,0 +1,163 @@
+# Proposal: `downloads.getFileHash()`
+
+**Summary**
+
+Allow extensions to query hashes of contents of downloaded files
+programmatically.
+
+**Document Metadata**
+
+**Author:** @bershanskyi
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** @Rob--W
+
+**Created:** 2024-04-08
+
+**Related Issues:** [WECG GitHub issue 401](https://github.com/w3c/webextensions/issues/401)
+
+## Motivation
+
+### Objective
+
+This API will enable extensions to privately and efficiently scan user
+downloads for "Known Unsafe" contents and validate the integrity of "Known Safe"
+files without exposing the raw content of the files.
+
+#### Use Cases
+
+##### Filtering for unwanted downloads
+
+Many antivirus products detect previously seen "safe" and "unsafe" files by
+computing a file content hash and checking it against a set of known hashes.
+These checks require minimal effort on the client side, yet are powerful
+enough to limit the spread of known malware downloads since effectively every
+client gets "inoculated" against a particular download as soon as a single
+antivirus detects a malicious file.
+User-Agents may store file hashes even after the original file is deleted to
+facilitate tracing of malware exposure even if the file is deemed to be
+dangerous after the user initially downloads it.
+
+##### Download integrity assurance
+
+Files can get damaged while they are shared on a file hosting server or NAS.
+While some file formats explicitly support integrity checks like embedded
+checksums or even cryptographic signatures, other files do not. This API
+would allow file managers or extensions to programmatically check the
+integrity of downloaded files if the file source elects to provide them.
+
+### Known Consumers
+
+None at the moment.
+
+## Specification
+
+### Schema
+
+
+``` idl
+namespace downloads {
+    callback GetFileHashCallback = void(DOMString hash);
+
+    [supportsPromises, permissions="downloads.getFileHash"]
+    static void getFileHash(
+        long downloadId,
+        [legalValues=("SHA-256")] DOMString algorithm,
+        GetFileHashCallback callback);
+}
+```
+
+``` idl
+namespace downloads {
+  enum FileHashAlgorithm {
+    SHA256 = "SHA-256"
+  }
+
+  getFileHash: (
+    downloadId: number,
+    algorithm?: keyof downloads.FileHashAlgorithm)
+      => Promise<string>;
+}
+```
+
+### New Permissions
+
+|      Permission Added     | Suggested Warning |
+| ------------------------- | ----------------- |
+| `"downloads.getFileHash"` | No warning        |
+
+The new `"downloads.getFileHash"` permission is optional and is honored only if
```

Comment:
> is honored only if extension has `"downloads"` permission as well

What was the thinking here? I think in general these sub-permissions work without the parent permissions so this might be inconsistent with downloads.open for example.

Reply:
Sub-permissions work a bit differently across browsers: Firefox requires both `"downloads"` and `"downloads.open"` to call `downloads.open()` method. I do not have a strong opinion on this. Personally, I would be more happy with the way Chrome does it (since more fine-grained permissions carry greater agency for the user/user-agent). Would be nice to discuss this during the call today.

Reply:
I tested in Chrome and it behaves like Firefox: The `downloads` permission is required for access to the `downloads` namespace. The `downloads.open` permission has a separate permission warning, and unlocks the `downloads.open` method.

After having thought this through a bit, I think that the behavior as described is good enough. It is not only consistent with what we already have, the most compelling point is that the `downloads.getFileHash` method is of no use without access to the rest of the `downloads` API. In particular, the `downloadId` needed to identify specific downloads can only be obtained through the `downloads` permission.

Reply:
Thanks Rob, it seems like I was mistaken - I confirmed with Devlin and your description matches the intended behavior.

On the existence of the permission in general, we want to add permissions for new namespaces but not necessarily new APIs - as an example we have a `userScripts` permission but not one for each of the individual methods. The one exception is if the API is particularly more dangerous, but given the host permission requirement, Devlin and I don't feel that it meets that bar. Do you feel strongly on having a permission here / have a case for why this is significantly more dangerous than host permissions? If so I think there will be some additional work to align.

Reply:
This permission predates my suggestion of relying on host permissions. With the use of host permissions, the only remaining value of the permission is to audit API usage more easily. The API is not distinctive enough (e.g. compared to fetch/XHR) to really need that though.

Reply:
Ok, great - in that case it seems like we're aligned on removing the permission. @bershanskiy, does that sound good to you?

Reply:
FYI @xeenon

Reply:
I removed `"downloads.getFileHash"` permission [here](https://github.com/w3c/webextensions/compare/57c0ca994058f858c7d78eec6bff02c95f5a098a..591f58d1b321e9bc9682696d64e2df4bcb75c4c1). Please mark as "resolved" if this works OK.